File: test-fixtures/python/bad_code.py
"""Deliberately bad code. Fixture for testing the quality-report skill.

Planted problems:
- process_data: complexity > 10, length > 50 lines, nesting > 4,
  and a comment that describes HOW instead of WHAT.
- build_report: 7 arguments.
- summarize_sales / summarize_refunds: duplicated >= 5-line block.
- legacy_export: dead (never called) and uses a bare except.
"""


def process_data(data):
    # Loop through the items, use nested ifs to inspect each field,
    # and append the transformed value to the results list.
    results = []
    for item in data:
        if item is not None:
            if isinstance(item, dict):
                if "value" in item:
                    if item["value"] > 0:
                        if item["value"] < 100:
                            results.append(item["value"] * 2)
                        else:
                            results.append(100)
                    else:
                        results.append(0)
                elif "name" in item:
                    if item["name"]:
                        results.append(len(item["name"]))
                    else:
                        results.append(-1)
                elif "flag" in item:
                    if item["flag"] is True:
                        results.append(1)
                    else:
                        results.append(0)
                else:
                    results.append(None)
            elif isinstance(item, list):
                if len(item) > 10:
                    results.append(sum(item[:10]))
                elif len(item) > 5:
                    results.append(sum(item[:5]))
                elif len(item) > 0:
                    results.append(item[0])
                else:
                    results.append(0)
            elif isinstance(item, str):
                if item.startswith("A"):
                    results.append(1)
                elif item.startswith("B"):
                    results.append(2)
                elif item.startswith("C"):
                    results.append(3)
                else:
                    results.append(0)
            elif isinstance(item, int):
                if item % 2 == 0:
                    results.append(item // 2)
                else:
                    results.append(item * 3 + 1)
            else:
                results.append(None)
    return results


def build_report(name, date, author, title, status, priority, category):
    """Build a report dictionary from its parts."""
    return {
        "name": name,
        "date": date,
        "author": author,
        "title": title,
        "status": status,
        "priority": priority,
        "category": category,
    }


def summarize_sales(records):
    """Return the average positive sale amount."""
    total = 0
    count = 0
    for record in records:
        if record["amount"] > 0:
            total += record["amount"]
            count += 1
    return total / count if count else 0


def summarize_refunds(records):
    """Return the average positive refund amount."""
    total = 0
    count = 0
    for record in records:
        if record["amount"] > 0:
            total += record["amount"]
            count += 1
    return total / count if count else 0


def legacy_export(records):
    """Export records as strings."""
    try:
        return [str(record) for record in records]
    except:
        return []


if __name__ == "__main__":
    sample = [{"value": 5}, [1, 2, 3], "Apple", 7, None]
    assert process_data(sample) == [10, 1, 1, 22]
    assert summarize_sales([{"amount": 10}, {"amount": -2}]) == 10
    assert summarize_refunds([{"amount": 4}, {"amount": 6}]) == 5
    print("fixture self-check ok")

File: test-fixtures/javascript/bad_code.ts
/**
 * Deliberately bad TypeScript. Fixture for testing the quality-report skill.
 *
 * Planted problems:
 * - Ledger.classify: complexity > 10, nesting > 4, and a comment that
 *   describes HOW instead of WHAT.
 * - Ledger.constructor: 6 arguments.
 * - totalSales / totalRefunds: duplicated >= 5-line block.
 * - swallow: empty catch, an unused variable, and no comment of its own.
 */

export interface Entry {
  amount: number;
  label?: string;
}

export class Ledger {
  // Assign each constructor argument to an instance field one at a time.
  constructor(
    public name: string,
    public owner: string,
    public currency: string,
    public opened: string,
    public closed: string,
    public region: string,
  ) {}

  // Walk the entry through nested ifs on amount and label, returning the
  // first bucket name that matches.
  classify(entry: Entry): string {
    if (entry.amount > 0) {
      if (entry.label) {
        if (entry.label.length > 3) {
          if (entry.label.startsWith('x')) {
            if (entry.amount > 100) {
              return 'big-x';
            }
            return 'x';
          }
          return 'long';
        }
        return 'short';
      }
      return entry.amount > 50 ? 'large' : 'small';
    }
    if (entry.amount < 0) {
      return entry.label ?? 'refund';
    }
    if (Number.isNaN(entry.amount) || entry.label === undefined) {
      return 'unknown';
    }
    return 'zero';
  }
}

/** Return the average positive sale amount. */
export function totalSales(entries: Entry[]): number {
  let total = 0;
  let count = 0;
  for (const entry of entries) {
    if (entry.amount > 0) {
      total += entry.amount;
      count += 1;
    }
  }
  return count ? total / count : 0;
}

/** Return the average positive refund amount. */
export function totalRefunds(entries: Entry[]): number {
  let total = 0;
  let count = 0;
  for (const entry of entries) {
    if (entry.amount > 0) {
      total += entry.amount;
      count += 1;
    }
  }
  return count ? total / count : 0;
}

export function swallow(entries: Entry[]): string[] {
  const unusedTotal = 0;
  try {
    return entries.map((entry) => String(entry.amount));
  } catch (err) {
  }
  return [];
}
